### ISSUE TITLE
PPO ratio explosion and logprob of the behavior policy

### ISSUE THREAD

**Issue by vwxyzjn** [lvwerra/trl#454]
# Issue Description

I noticed an issue with the `logprob` and `ratio` calculation. When printing out the `ratio` in the sentiment analysis example, I get diverging ratios ([tracked logs](https://wandb.ai/costa-huang/trl/runs/mw14ilad/logs?workspace=user-costa-huang)). This should be incorrect , however, because the `ratio` should be all exactly 1s during the first update eoch.
https://github.com/lvwerra/trl/blob/7705daa672f9264d7b5b789b6a1fd6b1cff03a58/trl/trainer/ppo_trainer.py#L1037

```
   6 ratio.mean() tensor(1., device='cuda:0', grad_fn=<MeanBackward0>)
   7 ratio.mean() tensor(0.9986, device='cuda:0', grad_fn=<MeanBackward0>)
   8 ratio.mean() tensor(0.9995, device='cuda:0', grad_fn=<MeanBackward0>)
   9 ratio.mean() tensor(1.0109, device='cuda:0', grad_fn=<MeanBackward0>)
  10 ratio.mean() tensor(1.0214, device='cuda:0', grad_fn=<MeanBackward0>)
  11 ratio.mean() tensor(1.0187, device='cuda:0', grad_fn=<MeanBackward0>)
  12 ratio.mean() tensor(1.0211, device='cuda:0', grad_fn=<MeanBackward0>)
  13 ratio.mean() tensor(1.0299, device='cuda:0', grad_fn=<MeanBackward0>)
  14 ratio.mean() tensor(1.0259, device='cuda:0', grad_fn=<MeanBackward0>)
  15 ratio.mean() tensor(1.0321, device='cuda:0', grad_fn=<MeanBackward0>)
  16 ratio.mean() tensor(1.0232, device='cuda:0', grad_fn=<MeanBackward0>)
  17 ratio.mean() tensor(1.0351, device='cuda:0', grad_fn=<MeanBackward0>)
...
2414 ratio.mean() tensor(0.9862, device='cuda:0', grad_fn=<MeanBackward0>)
2415 ratio.mean() tensor(0.9681, device='cuda:0', grad_fn=<MeanBackward0>)
2416 37it [05:52,  7.92s/it]/fsx/costa/trl/trl/trainer/ppo_trainer.py:1104: UserWarning: KL divergence is starting to become negative: -175.58 - this might be a precursor for failed training. sometimes this happens because the generation kwargs are not correctly set. Please make sure that the generation kwargs are set correctly, or review your training hyperparameters.
2417 ratio.mean() tensor(1., device='cuda:0', grad_fn=<MeanBackward0>)
2418 ratio.mean() tensor(1.4301, device='cuda:0', grad_fn=<MeanBackward0>)
2419 ratio.mean() tensor(1.4391, device='cuda:0', grad_fn=<MeanBackward0>)
2420 ratio.mean() tensor(0.9457, device='cuda:0', grad_fn=<MeanBackward0>)
2421 ratio.mean() tensor(0.9582, device='cuda:0', grad_fn=<MeanBackward0>)
```

# Expected behavior
In the original code base, PPO calculates three logprobs during the rollout ([lm_human_preferences/train_policy.py#L285](https://github.com/openai/lm-human-preferences/blob/cbfd210bb8b08f6bc5c26878c10984b90f516c66/lm_human_preferences/train_policy.py#L285), [lm_human_preferences/policy.py#L97-L112](https://github.com/openai/lm-human-preferences/blob/cbfd210bb8b08f6bc5c26878c10984b90f516c66/lm_human_preferences/policy.py#L97-L112)):
* logprob of the behavior policy ($\pi_{\theta_{old}}$ in the PPO paper, $\theta_{old}$ is the vector of policy parameters before the update) 
* logprob of the target/current policy ($\pi_{\theta}$ in the PPO paper, this is a policy we are interested in optimizing)
* logprob of the reference policy (for RLHF and KL penalty for the rewards)

One useful debugging variable is `ratio = torch.exp(logprobs - old_logprobs)`. While the values of the ratio can vary, during the first PPO update epoch `ratio` should be all 1s. This is because $\pi_{\theta_{old}} = \pi_{\theta}$ in the first epoch where no policy updates have happened. 

When printing out CLeanRL's PPO's [ratio](https://github.com/vwxyzjn/cleanrl/blob/f155a89cc12805a2da70958a3fcae8d3d11682ce/cleanrl/ppo.py#L248) with `python ppo.py --cuda False --num-minibatches 1` (here we disable minibatches like TRL), we get

```
epoch=0, ratio=1.0
epoch=1, ratio=0.9996373653411865
epoch=2, ratio=0.9992523193359375
epoch=3, ratio=0.9988474249839783
...
epoch=0, ratio=1.0
epoch=1, ratio=1.00007164478302
epoch=2, ratio=1.000141978263855
epoch=3, ratio=1.0002108812332153
```

<img width="974" alt="image" src="https://github.com/lvwerra/trl/assets/5555347/320e450c-9d25-4bab-9331-8a7d9f6a6709">

## Actual behavior

In TRL, we used `ppo_epochs = 4 `, which means the `ratio` should be all 1s every 4 prints, but it was not the case.

This may be the root cause of all stability issues like #423  #355 #235

**Comment by vwxyzjn** [comment]
After digging into a bit, the different ratios shown in the output mainly come from a lot of mini-batches. As shown below, the ratio becomes 1s when epoch is zero again. Note that we are doing 4 * 16 = 64 gradient updates per batch of data. uses 1 minibatch ([lm_human_preferences/train_policy.py#L50](https://github.com/openai/lm-human-preferences/blob/cbfd210bb8b08f6bc5c26878c10984b90f516c66/lm_human_preferences/train_policy.py#L50)). Maybe this is tangentially related to the issue.

```
epoch=0, i=0, ratio=1.0
epoch=0, i=1, ratio=1.0023666620254517
epoch=0, i=2, ratio=1.0039550065994263
epoch=0, i=3, ratio=1.0023411512374878
epoch=0, i=4, ratio=1.0051848888397217
epoch=0, i=5, ratio=1.009121298789978
epoch=0, i=6, ratio=1.0077825784683228
epoch=0, i=7, ratio=1.0123776197433472
epoch=0, i=8, ratio=1.0179272890090942
epoch=0, i=9, ratio=1.0296952724456787
epoch=0, i=10, ratio=1.0253453254699707
epoch=0, i=11, ratio=1.0136613845825195
epoch=0, i=12, ratio=1.0205295085906982
epoch=0, i=13, ratio=1.026404857635498
epoch=0, i=14, ratio=1.0109573602676392
epoch=0, i=15, ratio=1.0167485475540161
epoch=1, i=0, ratio=1.0325846672058105
epoch=1, i=1, ratio=1.100885272026062
epoch=1, i=2, ratio=1.0783733129501343
epoch=1, i=3, ratio=1.042301058769226
epoch=1, i=4, ratio=1.034325122833252
epoch=1, i=5, ratio=1.0306745767593384
epoch=1, i=6, ratio=1.0398730039596558
epoch=1, i=7, ratio=0.9988977313041687
epoch=1, i=8, ratio=1.0230234861373901
epoch=1, i=9, ratio=1.0378843545913696
epoch=1, i=10, ratio=0.9869514107704163
epoch=1, i=11, ratio=1.01687490940094
epoch=1, i=12, ratio=1.0925251245498657
epoch=1, i=13, ratio=1.033782958984375
epoch=1, i=14, ratio=1.0568681955337524
epoch=1, i=15, ratio=1.0245441198349
epoch=2, i=0, ratio=1.0052833557128906
epoch=2, i=1, ratio=0.9887069463729858
epoch=2, i=2, ratio=0.9870575666427612
epoch=2, i=3, ratio=0.9837588667869568
epoch=2, i=4, ratio=1.0077781677246094
epoch=2, i=5, ratio=0.9774166345596313
epoch=2, i=6, ratio=1.1206282377243042
epoch=2, i=7, ratio=0.9610921740531921
epoch=2, i=8, ratio=1.0360815525054932
epoch=2, i=9, ratio=1.1025735139846802
epoch=2, i=10, ratio=0.9839668273925781
epoch=2, i=11, ratio=1.0954787731170654
epoch=2, i=12, ratio=0.947047233581543
epoch=2, i=13, ratio=1.063216209411621
epoch=2, i=14, ratio=1.0657410621643066
epoch=2, i=15, ratio=1.0122908353805542
epoch=3, i=0, ratio=1.1595962047576904
epoch=3, i=1, ratio=1.0290977954864502
epoch=3, i=2, ratio=1.0603705644607544
epoch=3, i=3, ratio=0.9300947785377502
epoch=3, i=4, ratio=0.9222888350486755
epoch=3, i=5, ratio=1.0095583200454712
epoch=3, i=6, ratio=1.0408023595809937
epoch=3, i=7, ratio=1.0108882188796997
epoch=3, i=8, ratio=1.0525754690170288
epoch=3, i=9, ratio=1.044208288192749
epoch=3, i=10, ratio=1.0530775785446167
epoch=3, i=11, ratio=1.0434566736221313
epoch=3, i=12, ratio=0.9507423043251038
epoch=3, i=13, ratio=1.03242826461792
epoch=3, i=14, ratio=1.1840221881866455
epoch=3, i=15, ratio=1.0326350927352905
epoch=0, i=0, ratio=1.0
epoch=0, i=1, ratio=1.0003670454025269
epoch=0, i=2, ratio=1.0054229497909546
epoch=0, i=3, ratio=1.009329080581665
epoch=0, i=4, ratio=1.0136653184890747
epoch=0, i=5, ratio=1.0142427682876587
epoch=0, i=6, ratio=1.0175950527191162
epoch=0, i=7, ratio=1.0130765438079834
epoch=0, i=8, ratio=1.0121607780456543
epoch=0, i=9, ratio=1.007230281829834
epoch=0, i=10, ratio=1.0159568786621094
epoch=0, i=11, ratio=1.0155819654464722
epoch=0, i=12, ratio=1.0219719409942627
epoch=0, i=13, ratio=1.0138139724731445
epoch=0, i=14, ratio=1.0263171195983887
epoch=0, i=15, ratio=0.9543671607971191
epoch=1, i=0, ratio=0.9743574261665344
epoch=1, i=1, ratio=0.986846387386322
epoch=1, i=2, ratio=0.8345062136650085
epoch=1, i=3, ratio=0.912946343421936
epoch=1, i=4, ratio=0.8850008845329285
epoch=1, i=5, ratio=0.8190274834632874
epoch=1, i=6, ratio=0.7437756657600403
epoch=1, i=7, ratio=0.8041974902153015
epoch=1, i=8, ratio=0.7328497767448425
epoch=1, i=9, ratio=0.9503921270370483
epoch=1, i=10, ratio=0.8375219106674194
epoch=1, i=11, ratio=0.7887753248214722
epoch=1, i=12, ratio=0.812556803226471
epoch=1, i=13, ratio=0.7109256982803345
epoch=1, i=14, ratio=0.8298351168632507
epoch=1, i=15, ratio=0.7963868975639343
epoch=2, i=0, ratio=0.774953305721283
epoch=2, i=1, ratio=0.755143404006958
epoch=2, i=2, ratio=0.8826191425323486
epoch=2, i=3, ratio=0.8266997933387756
epoch=2, i=4, ratio=0.7973554730415344
epoch=2, i=5, ratio=0.7650317549705505
epoch=2, i=6, ratio=0.8091626763343811
epoch=2, i=7, ratio=0.766830563545227
epoch=2, i=8, ratio=0.7297632098197937
epoch=2, i=9, ratio=0.8460482954978943
epoch=2, i=10, ratio=0.8089452385902405
epoch=2, i=11, ratio=0.8457133173942566
epoch=2, i=12, ratio=0.7562631964683533
epoch=2, i=13, ratio=0.7832910418510437
epoch=2, i=14, ratio=0.7803390026092529
epoch=2, i=15, ratio=0.8795580267906189
epoch=3, i=0, ratio=0.8717175722122192
epoch=3, i=1, ratio=0.8007976412773132
epoch=3, i=2, ratio=0.8370493054389954
epoch=3, i=3, ratio=0.7741546034812927
epoch=3, i=4, ratio=0.8467444777488708
epoch=3, i=5, ratio=0.7997295260429382
epoch=3, i=6, ratio=0.7450951933860779
epoch=3, i=7, ratio=0.7487857341766357
epoch=3, i=8, ratio=0.8477750420570374
epoch=3, i=9, ratio=0.7335952520370483
epoch=3, i=10, ratio=0.8565078973770142
epoch=3, i=11, ratio=0.8273767828941345
epoch=3, i=12, ratio=0.8224087953567505
epoch=3, i=13, ratio=0.8002591729164124
epoch=3, i=14, ratio=0.6742428541183472
epoch=3, i=15, ratio=0.8572400808334351
```
